충남대 BE_좌진우 2단계 - 주문하기

jinwoojwa opened the pull request:
## Step 2 - 주문 하기

🧾 이전 피드백 반영 사항

- 로그인 이후 클라이언트가 사용자 정보를 활용할 수 있도록, 본인 식별을 위한 값들을 응답에 포함했습니다.
  + KakaoLoginResponseDto를 생성하여 JWT와 카카오 토큰 정보들을 반환하도록 구현했습니다.

<br>

📚 구현 사항

1️⃣ 주문하기(Order) 기능 구현
- 사용자가 상품 옵션과 수량, 메시지를 입력해 주문을 할 수 있도록 하는 API 구현
- 주문 생성 시:
  + 옵션 수량 차감
  + 해당 상품이 위시리스트에 있으면 제거
  + 카카오톡 메시지 나에게 보내기 API를 활용해 주문 정보를 메시지로 전송
- 메시지 전송은 `KakaoMessageClient`를 통해 수행
- 메시지 템플릿은 `KakaoMessageTemplateMaker`에서 구성

2️⃣ 카카오톡 메시지 보내기 API
- `KakaoMessageClient`에서 처리하도록 구현
- 주문 템플릿은 `기본 템플릿` 중 `텍스트 템플릿`을 사용하여 구현

<br>

💻 테스트

<img width="222" height="232" alt="image" src="https://github.com/user-attachments/assets/e81f8361-8178-4f9f-a213-5f831197dbee" />

---

📌 추가 구현 계획

구현 사항이 너무 많아지고 시간이 오래 걸릴 것 같아 먼저 PR 드렸습니다. 아래 내용 추가할 계획입니다.

- `주문하기` 관련 API 호출 테스트 추가
- 토큰 만료 및 재발급 문제 해결
- 그 외 예외 처리 코드 추가

jinwoojwa committed:
feat: UserKakaoToken 클래스 생성 및 Member와 관계 설정

jinwoojwa committed:
feat: KakaoLoginResponseDto 추가

- 로그인 후 JWT, 액세스 토큰을 반환하기 위해 DTO 추가

jinwoojwa committed:
refactor: 패키지명 auth에서 oauth로 변경

jinwoojwa committed:
docs(README): step2 기능 요구 사항 및 세부 구현 사항 정리

jinwoojwa committed:
refactor: Kakao OAuth와 메시지 API 클라이언트 분리 및 RestClient 주입 방식 개선

- Kakao API와 인증 서버 요청을 분리하기 위해 RestClient를 @Qualifier로 등록
  - kakaoApiClient: https://kapi.kakao.com
  - kakaoAuthClient: https://kauth.kakao.com
- KakaoOAuthClient에서 RestClient 인스턴스를 kakaoAuthClient로 교체
- application.properties에서 관련 경로 설정
- KakaoOauthProperties에 messageUrl 필드 추가

jinwoojwa committed:
feat: 카카오 메시지 템플릿 생성 클래스 구현

jinwoojwa committed:
feat: 카카오 메시지 API 클라이언트 구현

jinwoojwa committed:
feat: 주문하기 기능 구현

- 상품 옵션 및 수량을 입력받아 주문 생성
- 재고 차감 및 위시리스트에서 자동 제거
- 주문 요청/응답용 DTO 추가
- 주문 시간 및 사용자 메시지 포함

jinwoojwa committed:
옵션 예외 처리 클래스 추가 및 핸들러 코드 추가

jinwoojwa committed:
fix: 카카오 OAuth 클라이언트 URL 및 RestClient 사용 방식 수정

- kakaoAuthClient의 baseUrl과 tokenUrl 경로 문제 해결
- 잘못된 API 호출로 인한 404 에러 수정

leaf-upper commented on README.md:
```diff
@@ -66,4 +66,67 @@
+
+---
+
+#### 기능 요구 사항
```
Readme를 너무 잘 정리해주셨네요
👍 👍

leaf-upper commented on src/main/java/gift/oauth/client/KakaoMessageClient.java:
```diff
@@ -0,0 +1,48 @@
+    private final KakaoMessageTemplateMaker templateMaker;
+
+    public KakaoMessageClient(
+            @Qualifier("kakaoApiClient") RestClient restClient,
```
Qualifier를 사용하셔도 좋지만,
RestClient를 빈으로 등록하셨다면, BeanName을 통해서도 빈주입을 받으실 수 있습니다.
`RestClient kakaoApiClient` 요런식으로도 빈주입이 가능하지않을까요?

leaf-upper replied:
위 코멘트에도 불구하고 굉장히 깔끔한 Client코드네요 👍 👍 
이런식으로 코드를 작성해주시면 좋을것 같아요

leaf-upper commented on src/main/java/gift/oauth/client/KakaoMessageClient.java:
```diff
@@ -0,0 +1,48 @@
+            throw new IllegalArgumentException("❗ 메시지 인코딩 실패", e);
+        }
+
+        kakaoRestClient.post()
```
외부서버와 통신하는 RestClient는 서버오류, 네트워크 오류 등으로 장애가 발생할 수 있습니다.
이런 상황에 대처하기 위해서 적절한 예외처리가 필요할거 같아요.

leaf-upper commented on src/main/java/gift/oauth/client/KakaoOAuthClient.java:
```diff
@@ -27,7 +33,7 @@ public String getAuthorizeUrl() {
         String body = kakaoProps.getTokenRequestBody(code);
 
-        return restClient.post()
+        return kakaoAuthClient.post()
```
위에서 코멘트 드린부분과 동일합니다. RestClient 통신 코드에 대해서 적절한 예외처리를 해주시면 좋을것 같아요

leaf-upper commented on src/main/java/gift/oauth/config/KakaoRestClientConfig.java:
```diff
@@ -0,0 +1,30 @@
+    @Bean
+    @Qualifier("kakaoApiClient") // https://kapi.kakao.com
+    public RestClient kakaoApiClient() {
+        return RestClient.builder()
```
Bean으로 등록해주신것도 좋고 Qualifier가 아니라, BeanName을 지정하시는 방식도 좋을것 같아요.
BeanName이 따로 없는경우, 메서드명이 BeanName으로 포함됩니다.

leaf-upper commented on src/main/java/gift/oauth/config/KakaoRestClientConfig.java:
```diff
@@ -0,0 +1,30 @@
+    public RestClient kakaoApiClient() {
+        return RestClient.builder()
+                .baseUrl("https://kapi.kakao.com")
+                .defaultHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_FORM_URLENCODED_VALUE)
```
RestClient.Builder를 사용하셨네요. RestClient.Builder는 자동구성에 의해 자동으로 빈으로 등록되니 해당 Bean을 이용하여 RestClient를 생성하는것도 방법입니다.

leaf-upper replied:
여기서 errorHandler를 등록해주셔도 좋을것 같아요.

leaf-upper commented on src/main/java/gift/oauth/service/DefaultKakaoAuthService.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    @Transactional
+    public KakaoLoginResponseDto loginWithKakao(String code) {
```
👍 👍 굉장히 깔끔한 코드인것 같아요~

leaf-upper commented on src/main/java/gift/oauth/util/KakaoMessageTemplateMaker.java:
```diff
@@ -0,0 +1,34 @@
+        this.objectMapper = objectMapper;
+    }
+
+    public String makeMessageTemplate(String message) {
```
template을 Map 타입을 통해 만드셔도 좋지만 클래스를 별도로 선언해서 만드는건 어떠실까요?

leaf-upper commented on src/main/java/gift/order/service/DefaultOrderService.java:
```diff
@@ -0,0 +1,100 @@
+        Order order = new Order(member, option, requestDto.quantity(), requestDto.message(), LocalDateTime.now());
+        Order savedOrder = orderRepository.save(order);
+
+        // 카카오톡 메시지 전송
```
일반적으로 메세지 전송은 주문과 별도의 쓰레드에서 비동기로 동작하는 경우가 많아요~
주문 완료 이후 메세지를 전송하도록 하는건 어떨까요?
1. `ApplicationEventPublisher` Spring Event 방식으로 주문 완료 이벤트 생성
2. 주문 완료 이벤트를 구독하여 사용자에게 메세지 전송
이때 이벤트는 주문이라는 데이터가 DB에 커밋 완료된 이후로 가면 좋을것 같아요
관련하여 `@TransactionEventListener` `@Async` 등의 애너테이션을 한번 참고해보시면 좋을것 같습니다.

leaf-upper commented on src/main/java/gift/order/service/DefaultOrderService.java:
```diff
@@ -0,0 +1,100 @@
+    }
+
+    private String buildMessage(Order order) {
+        return """
```
이런 템플릿성의 메세지도 Property로 관리하면 좋을것 같아요~

leaf-upper reviewed (changes requested):
안녕하세요 진우님~
Step2 고생많으셨습니다.
제가 중요하다고 생각하는 포인트에 대해서만 몇가지 코멘트를 남겨드렸어요
확인해보시고 수정 PR 주세요~
고생하셨습니다!

jinwoojwa committed:
refactor: 불필요한 Qualifier 제거

jinwoojwa committed:
feat: RestClient에 타임아웃 설정 적용 및 클라이언트 코드 네트워크 예외 처리

jinwoojwa committed:
refactor: 불필요한 Qualifier 제거

jinwoojwa committed:
refactor: RestClient 예외 처리 공통화 및 중복 코드 제거

- defaultStatusHandler를 RestClient 설정에 공통 적용하여 예외 처리
- 각 요청 메서드(requestToken, getUserInfo) 내 onStatus 중복 처리 제거

jinwoojwa committed:
refactor: 메시지 템플릿 생성을 Map에서 record 기반으로 리팩토링

jinwoojwa committed:
refactor: 주문 메시지 템플릿을 application.properties로 분리

jinwoojwa committed:
refactor: 카카오 메시지 템플릿 JSON 직렬화를 위한 @JsonProperty 추가

jinwoojwa committed:
chore: 주문 메시지 템플릿 및 한글 인코딩 설정 추가

jinwoojwa committed:
feat: 주문 메시지 전송을 이벤트 기반 비동기 처리로 분리

- 주문 완료 시 OrderEvent 발행
- @TransactionalEventListener로 메시지 전송 로직 비동기 분리
- KakaoMessageClient를 통해 카카오톡 메시지 전송
- Spring Boot 애플리케이션에 @EnableAsync 활성화

jinwoojwa commented:
🧾 변경 사항

- 불필요한 Qualifier 제거
- 클라이언트 코드들에 네트워크 예외 처리를 추가
- 메시지 템플릿 생성 코드를 record를 사용하도록 변경
- 주문 후 메시지 전송을 이벤트 기반의 비동기 처리로 분리
  + 주문 완료 시 OrderEvent를 발행하여 리스너에서 받아 처리하도록 구현

<br>

테스트 코드 및 일부 추가 기능 구현을 하려고 했으나, 미션의 난이도가 어려워 기본적인 미션 요구 사항을 구현하는 데 급급했던 것 같습니다. 계속해서 추가, 보완하도록 하겠습니다.

leaf-upper commented on src/main/java/gift/Application.java:
```diff
@@ -2,8 +2,10 @@
+import org.springframework.scheduling.annotation.EnableAsync;
 
 @SpringBootApplication
+@EnableAsync
```
👍 비동기 처리를 위해서 추가해주셨군요

leaf-upper commented on src/main/java/gift/oauth/util/KakaoMessageTemplateMaker.java:
```diff
@@ -0,0 +1,36 @@
+import org.springframework.stereotype.Component;
+
+@Component
+public class KakaoMessageTemplateMaker {
```
RestClient에서 body에 Object를 넣으면 Converter (MappingJacksonConverter)에 의해서 
자동으로 JSON으로 바뀌긴 할거에요~

leaf-upper reviewed (approved):
👍 안녕하세요 진우님 스텝2 고생많으셨습니다.
몇가지 코멘트만 첨부했고, 해당 스텝은 어프로브 드릴게요!
고생하셨습니다!
